odoh.go

```diff
@@ -389,44 +336,55 @@ func CreateDefaultKeyPair() (ObliviousDoHKeyPair, error) {
 
 type QueryContext struct {
 	odohSecret []byte
-	suite      hpke.CipherSuite
+	suite      hpke.Suite
 	query      []byte
 	publicKey  ObliviousDoHConfigContents
 }
 
 func (c QueryContext) DecryptResponse(message ObliviousDNSMessage) ([]byte, error) {
 	aad := append([]byte{byte(ResponseType)}, []byte{0x00, 0x00}...) // 0-length encoded KeyID
 
-	odohPRK := c.suite.KDF.Extract(c.query, c.odohSecret)
-	key := c.suite.KDF.Expand(odohPRK, []byte(ODOH_LABEL_KEY), c.suite.AEAD.KeySize())
-	nonce := c.suite.KDF.Expand(odohPRK, []byte(ODOH_LABEL_NONCE), c.suite.AEAD.NonceSize())
+	h := c.suite.KdfID.Hash()
+	odohPRK := hkdf.Extract(h.New, c.query, c.odohSecret)
+	key := make([]byte, c.suite.AeadID.KeySize())
+	reader := hkdf.Expand(h.New, odohPRK, []byte(ODOH_LABEL_KEY))
+	io.ReadFull(reader, key)
 
-	aead, err := c.suite.AEAD.New(key)
+	aead, err := c.suite.AeadID.New(key)
 	if err != nil {
 		return nil, err
 	}
+	nonce := make([]byte, aead.NonceSize())
+	reader = hkdf.Expand(h.New, odohPRK, []byte(ODOH_LABEL_NONCE))
+	io.ReadFull(reader, nonce)
 
 	return aead.Open(nil, nonce, message.EncryptedMessage, aad)
 }
 
 type ResponseContext struct {
 	query      []byte
-	suite      hpke.CipherSuite
+	suite      hpke.Suite
 	odohSecret []byte
 }
 
 func (c ResponseContext) EncryptResponse(response *ObliviousDNSResponse) (ObliviousDNSMessage, error) {
 	aad := append([]byte{byte(ResponseType)}, []byte{0x00, 0x00}...) // 0-length encoded KeyID
 
-	odohPRK := c.suite.KDF.Extract(c.query, c.odohSecret)
-	key := c.suite.KDF.Expand(odohPRK, []byte(ODOH_LABEL_KEY), c.suite.AEAD.KeySize())
-	nonce := c.suite.KDF.Expand(odohPRK, []byte(ODOH_LABEL_NONCE), c.suite.AEAD.NonceSize())
+	h := c.suite.KdfID.Hash()
+	odohPRK := hkdf.Extract(h.New, c.query, c.odohSecret)
+	key := make([]byte, c.suite.AeadID.KeySize())
```

Comment:
This pattern (allocate slice, create reader, fill buffer) is shared for key and nonce, and duplicated. Can we move it to a function that takes the PRK, label, and output length as input? (Note that if we revert as suggested above, we get that for free.)